kubetest2/kubetest2-gce/deployer/common.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployer
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"k8s.io/klog"
+	"sigs.k8s.io/boskos/client"
+	boskosCommon "sigs.k8s.io/boskos/common"
+)
+
+func (d *deployer) init() error {
+	if d.commonOptions.ShouldBuild() {
+		if err := d.verifyBuildFlags(); err != nil {
+			return fmt.Errorf("init failed to check build flags: %s", err)
+		}
+	}
+
+	if d.commonOptions.ShouldUp() {
+		if err := d.verifyUpFlags(); err != nil {
+			return fmt.Errorf("init failed to verify flags for up: %s", err)
+		}
+
+		if d.GCPProject == "" {
+			klog.Info("No GCP project provided, acquiring from Boskos")
+
+			if err := d.getProjectFromBoskos(); err != nil {
+				return fmt.Errorf("init failed to get project from boskos: %s", err)
+			}
+			klog.Infof("Got project %s from boskos", d.GCPProject)
+		}
+
+	}
+
+	if d.commonOptions.ShouldDown() {
+		if err := d.verifyDownFlags(); err != nil {
+			return fmt.Errorf("init failed to verify flags for down: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// getProjectFromBoskos creates a boskos client, acquires a gcp project
+// and starts a heartbeat goroutine to keep the project reserved
+func (d *deployer) getProjectFromBoskos() error {
+	boskos, err := client.NewClient(
+		os.Getenv("JOB_NAME")+"-kubetest2",
+		d.BoskosLocation,
+		"",
+		"",
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create boskos client: %s", err)
+	}
+
+	resourceType := "gce-project"
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
+
+	boskosProject, err := boskos.AcquireWait(ctx, resourceType, "free", "busy")
+	if err != nil {
+		return fmt.Errorf("failed to get a %s from boskos: %s", resourceType, err)
+	}
+	if boskosProject == nil {
+		return fmt.Errorf("boksos had no %s available", resourceType)
+	}
+
+	go func(c *client.Client, resource *boskosCommon.Resource) {
+		for range time.Tick(time.Minute * 5) {
+			if err := c.UpdateOne(resource.Name, "busy", nil); err != nil {
+				klog.Warningf("[Boskos] Update of %s failed with %v", resource.Name, err)
+			}
+		}
+	}(boskos, boskosProject)
+
+	d.boskos = boskos
```

Comment:
with foresight of not needing to persist this across up and down. let's not store these in memory anyway.
shouldn't be too bad constructing the client again if we keep the same name.

Reply:
If the name of the boskos resource is retained, can the client be completely thrown away? I'm not sure if the owner field of the boskos struct matters: https://github.com/kubernetes-sigs/boskos/tree/master/client

Reply:
yes we also need to use the same owner.

Reply:
I did some experimenting and ran into https://github.com/kubernetes-sigs/boskos/blob/master/client/client.go#L334

It seems like we won't be able to decouple up and down in the boskos case because we would have to persist boskos client state which is a private struct field.

Reply:
that's sad to know. not sure what the motivation was for having in-memory storage in the client.
/cc @ixdy